server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1439,12 +1439,15 @@ private void innerOpenEngineAndTranslog() throws IOException {
         assert recoveryState.getRecoverySource().expectEmptyRetentionLeases() == false || getRetentionLeases().leases().isEmpty()
             : "expected empty set of retention leases with recovery source [" + recoveryState.getRecoverySource()
             + "] but got " + getRetentionLeases();
-        trimUnsafeCommits();
+        // don't trim anything if we are going to open a read only engine.
+        if (engineFactory.isReadOnlyEngineFactory() == false) {
```

Comment:
I think we should add an assertion below that ensures that if it's returning true that it's a subclass of ReadOnlyEngine.